### ISSUE TITLE
Wrong computation when using 64bit literals.

### ISSUE THREAD

**Issue by MarekKnapek** [open-watcom/open-watcom-v2#1047]
Watcom compiler does not support the `long long` type introduced in `C99` and in `C++11`, but it supports the `__int64` type instead. Doing mathematical computation with variables of this type works correctly, but doing mathematical computation with integer literals or immediates of this type does not work correctly.

Consider two following examples. In the firs one I cast the literal one to `unsigned __int64` and do mathematical computation with it. In the second one I cast the literal one to `unsigned __int64`, save it into variable, and then do mathematical computation with it. The difference is, that in the second example I use one more variable, that should not be needed, but apparently, it is needed, Watcom behaves differently.

The example is computing 64bit * 64bit multiplication producing 128bit result (two times 64bit). The computation is `0x0123456701234567 * 0x89abcdef89abcdef == 0x009ca39dca878d6493392fefc94e4629` but Watcom produces `0x009ca39ff0f9ed4393392fefc94e4629`.

Steps to reproduce:
 - Download and install (full) latest version (2023-02-06).
 - Run IDE.
 - Menu -> File -> New Project -> Select some file (such as test.wpj) -> Open.
 - Keep target type as is (WinNT/Win95, console).
 - Press insert key to insert new source file, select test.c, close.
 - Menu -> Options -> C Compiler Switches -> 10 Memory Model and Processor Switches -> Pentium Register-based calling (-5r). This step is not needed for Watcom 11 downloaded from winworldpc website as this calling convention is the default one for new projects.
 - Press F4 key to build the project.
 - Press CTRL+D to open debugger.
 - Press F8 or F10 repeatedly until you step thru all the computations and see the result in locals window.

Affected versions:
 - Watcom 11
 - Open Watcom v2 build from 2023-02-06

Not affected versions:
 - Visual Studio 2022
 - GCC for Windows 12.2.0

`test1.c`

```c
#define NDEBUG
#include <assert.h> /* assert */
#include <limits.h> /* CHAR_BIT */

#define num_type unsigned __int64

void mul(num_type const* x, num_type const* y, num_type* z, num_type* co)
{
	#define shift ((int)(sizeof(num_type) * CHAR_BIT / 2))
	#define mask ((num_type)(((num_type)(((num_type)(1)) << shift)) - ((num_type)(1))))

	num_type xlo;
	num_type xhi;
	num_type ylo;
	num_type yhi;
	num_type xylo;
	num_type xymi;
	num_type yxmi;
	num_type xyhi;

	assert(x);
	assert(y);
	assert(z);
	assert(co);
	assert(((sizeof(num_type) * CHAR_BIT) % 2) == 0);

	xlo = ((num_type)(*x & mask));
	xhi = ((num_type)(*x >> shift));
	ylo = ((num_type)(*y & mask));
	yhi = ((num_type)(*y >> shift));
	xylo = ((num_type)(xlo * ylo));
	xymi = ((num_type)(xlo * yhi));
	yxmi = ((num_type)(xhi * ylo));
	xyhi = ((num_type)(xhi * yhi));
	*z = ((num_type)(*x * *y));
	*co =
		((num_type)(
			((num_type)(
				((num_type)(
					((num_type)(xymi >> shift)) +
					((num_type)(yxmi >> shift))
				)) +
				xyhi
			)) +
			((num_type)(
				((num_type)(
					((num_type)(
						((num_type)(xymi & mask)) +
						((num_type)(yxmi & mask))
					)) +
					((num_type)(xylo >> shift))
				)) >>
				shift
			))
		));

	#undef shift
	#undef mask
}


int main()
{
	num_type a;
	num_type b;
	num_type c;
	num_type d;

	a = 0x01234567ul; a <<= 32; a |= 0x01234567ul;
	b = 0x89abcdeful; b <<= 32; b |= 0x89abcdeful;
	mul(&a, &b, &c, &d);
	/* correct: */
	/* a = 0x0123456701234567 */
	/* b = 0x89abcdef89abcdef */
	/* c = 0x93392fefc94e4629 */
	/* d = 0x009ca39dca878d64 */
	/* watcom: */
	/* a = 0x0123456701234567 */
	/* b = 0x89abcdef89abcdef */
	/* c = 0x93392fefc94e4629 */
	/* d = 0x009ca39ff0f9ed43 */
	return 0;
}
```

`test2.c`
```c
#define NDEBUG
#include <assert.h> /* assert */
#include <limits.h> /* CHAR_BIT */

#define num_type unsigned __int64

void mul(num_type const* x, num_type const* y, num_type* z, num_type* co)
{
	#define shift ((int)(sizeof(num_type) * CHAR_BIT / 2))
	#define mask ((num_type)(((num_type)(((num_type)(1)) << shift)) - ((num_type)(1))))

	num_type m;
	num_type xlo;
	num_type xhi;
	num_type ylo;
	num_type yhi;
	num_type xylo;
	num_type xymi;
	num_type yxmi;
	num_type xyhi;

	assert(x);
	assert(y);
	assert(z);
	assert(co);
	assert(((sizeof(num_type) * CHAR_BIT) % 2) == 0);

	m = mask;
	xlo = ((num_type)(*x & m));
	xhi = ((num_type)(*x >> shift));
	ylo = ((num_type)(*y & m));
	yhi = ((num_type)(*y >> shift));
	xylo = ((num_type)(xlo * ylo));
	xymi = ((num_type)(xlo * yhi));
	yxmi = ((num_type)(xhi * ylo));
	xyhi = ((num_type)(xhi * yhi));
	*z = ((num_type)(*x * *y));
	*co =
		((num_type)(
			((num_type)(
				((num_type)(
					((num_type)(xymi >> shift)) +
					((num_type)(yxmi >> shift))
				)) +
				xyhi
			)) +
			((num_type)(
				((num_type)(
					((num_type)(
						((num_type)(xymi & m)) +
						((num_type)(yxmi & m))
					)) +
					((num_type)(xylo >> shift))
				)) >>
				shift
			))
		));

	#undef shift
	#undef mask
}


int main()
{
	num_type a;
	num_type b;
	num_type c;
	num_type d;

	a = 0x01234567ul; a <<= 32; a |= 0x01234567ul;
	b = 0x89abcdeful; b <<= 32; b |= 0x89abcdeful;
	mul(&a, &b, &c, &d);
	/* correct: */
	/* a = 0x0123456701234567 */
	/* b = 0x89abcdef89abcdef */
	/* c = 0x93392fefc94e4629 */
	/* d = 0x009ca39dca878d64 */
	/* watcom: */
	/* a = 0x0123456701234567 */
	/* b = 0x89abcdef89abcdef */
	/* c = 0x93392fefc94e4629 */
	/* d = 0x009ca39dca878d64 */
	return 0;
}
```

**Comment by jmalak** [comment]
Thanks you for your bug report.
Please what compiler options you used?
Do you want to say that in IDE is different default setup, or you use old .wpj project file from Watcom 11 with OW 2?

A few notes. 
OW 2 support long long type in both modes C89 and C99.
Default for OW2 compiler is -6r.

**Comment by MarekKnapek** [comment]
Today, I am not able to reproduce this. Yesterday, I was able to reproduce this. Today, I have done additional testing: There is definitively a bug in Watcom 11 and there is no bug in Open Watcom 2.

But there is another problem: I have both W11 and OW2 installed on the same machine (in different folders, of course). I have not modified the environment variables during installation of both products. I have created *.bat file to modify the environment locally and then launch the IDE, one *.bat file for W11 and second different *.bat file file for OW2. I have both IDEs running at the same time (each having its own separate environment variables).

The symptoms: Sometimes one running instance of W affects the another running instance of W, sometimes W11 uses OW2 linker (WTF?), sometimes W IDE window freezes (not responding). So, the problem wit 64bit integers was, that OW2 was using compiler/linker from W11 and I was thus assuming there is a bug in OW2.

The cause: In the case of frozen IDE window, the main thread is blocked on waiting for single object Windows API waiting for ReadUpSem or ReadDoneSem semaphore, but this semaphore does not exist and no other application will create it, thus the IDE is frozen. I think this is bad design inherited from old Watcom IDE, assuming only one instance of W is running in machine (old Windows) or assuming only one instance of W is running in session (new Windows). Better design would be to create uniquely named semaphore each time before spawning new process, spawn new process with command line argument or with environment variable argument telling it the name of the semaphore to use, and then synchronize the two processes using that semaphore(s). There would be no room for other (third) process (for example from another instance of W) to interfere with such semaphore as there it is today. I am aware this is lot of work for little benefit (having two running instances of W at the same time seems to be rare).

**Comment by jmalak** [comment]
You are right, the design was done with assumption that only one instance is running at time.
The similar problem was with some "temporary" files which used fixed file name or numbered from some range.
Most of these issues were fixed but IDE is problematic part of OW distribution, it is not only one of problem.
There is backward compatibility problem because configuration files include some texts for options which were several time changed and therefore some compatibility breaks exists etc.
Anyway problem which you reported must be fixed for 16/32/64-bit Windows and OS/2 hosts, which are only platforms running OW IDE.
I will add it as enhancement request with hope that somebody will be interesting in fixing or if I will have a time, but I am very busy by more important bug fixes and new features, mainly 64-bit targets supports and C99 standard compliance.
